### ISSUE TITLE
Complete the implementation of storage::BucketMetadata

### ISSUE THREAD

**Issue by coryan** [GoogleCloudPlatform/google-cloud-cpp#537]
The `storage::BucketMetadata` class, currently under review in #535, is incomplete.  It does not support ACLs, or IAM, or CORS, or modifying attributes of the metadata. The first step in this issue should be to complete the list of missing features, and then break them down in smaller tasks.

**Comment by houglum** [comment]
Adding a note here, per the review in #686.  Currently, we have a class to represent metadata keys shared between both Objects and Buckets (`CommonMetadata`), and we also have classes to represent those resource types which also include the metadata attributes unique to those resource types (`BucketMetadata`, `ObjectMetadata`).  Here are the key sets that should be present in each of those classes for them to be considered complete:

And the common top-level keys used by both Objects and Buckets are:
```
>>> pprint.pprint(sorted(bktkeyset & objkeyset))
[u'acl',
 u'etag',
 u'id',
 u'kind',
 u'metageneration',
 u'name',
 u'owner',
 u'selfLink',
 u'storageClass',
 u'timeCreated',
 u'updated']
```

The keys unique to Object resources are:
```
>>> pprint.pprint(sorted(objkeyset - bktkeyset))
[u'bucket',
 u'cacheControl',
 u'componentCount',
 u'contentDisposition',
 u'contentEncoding',
 u'contentLanguage',
 u'contentType',
 u'crc32c',
 u'customerEncryption',
 u'generation',
 u'kmsKeyName',
 u'md5Hash',
 u'mediaLink',
 u'metadata',
 u'size',
 u'timeDeleted',
 u'timeStorageClassUpdated']
```

The keys unique to buckets are:
```
>>> pprint.pprint(sorted(bktkeyset - objkeyset))
[u'billing',
 u'cors',
 u'defaultObjectAcl',
 u'encryption',
 u'labels',
 u'lifecycle',
 u'location',
 u'logging',
 u'projectNumber',
 u'versioning',
 u'website']
```

**Comment by coryan** [comment]
I think this was fixed by #967.  All fields are implemented, including write-able fields. And optional fields using `google::cloud::internal::optional` too.
